File: DFS & BFS/LABYR1.py
import Queue
from collections import defaultdict
 
def bfs(row,col):
    
    visited = [[0]*c for _ in range(r)]
    sp = defaultdict(int)
    q = Queue.Queue(maxsize=c*r)
    q.put((row,col))
    visited[row][col] = 1
    sp[(row,col)] = 0
    
    dx = [1,-1,0,0]
    dy = [0,0,1,-1]
    
    while not q.empty():
        temp = q.get()
        i = temp[0]
        j = temp[1]
        
        for k in range(4):
            ti = i + dx[k]
            tj = j + dy[k]
            
            if (ti<0 or ti>=r) or (tj<0 or tj>=c) or visited[ti][tj] or mat[ti][tj]=="#":
                continue
            else:
                #print ti,tj
                visited[ti][tj] = True
                q.put((ti,tj))
                sp[(ti,tj)] = sp[(i,j)] + 1
        
        
            
    key = max(sp, key=lambda x: sp[x])
    #print key,sp[key]
    return key,sp[key]
 
t = int(raw_input())
 
for _ in range(t):
    c,r = map(int, raw_input().split())
    mat = []
    
    
    for _ in range(r):
        row = list(raw_input())
        mat.append(row)
        
    flag = False
    for row in range(r):
        for col in range(c):
            if (mat[row][col]=="."):
                key,value = bfs(row,col) 
                flag = True
                break
        if flag:
            break
    k,v = bfs(key[0],key[1])
    print "Maximum rope length is "+str(v)+"." 

File: DFS & BFS/LABYR1.cpp
using namespace std;

#include <iostream>
#include <queue>
#include <map>
#include <utility>

#define tr(c,i) for(typeof(c.begin()) i=c.begin(); i!=c.end(); i++)




int bfs(int i, int j,int r, int c,vector < vector <bool> >& visited, vector < vector <char> >& ch){
	//cout<<"now in bfs search"<<endl;
	map < pair <int,int>, int > sp;
	
	queue < pair <int,int> > q;
	q.push({i,j});
	visited[i][j] = true;
	sp[{i,j}] = 0;
	
	while (!q.empty()){
		pair <int,int> temp = q.front();
		q.pop();
		i = temp.first;
		j = temp.second;
		//cout<<"in queue "<<i<<" "<<j<<endl;
		
		if ((i-1 >=0)&&(!visited[i-1][j])&&(ch[i-1][j]=='.')){
			q.push({i-1,j});
			sp[{i-1,j}] = sp[{i,j}] + 1;
			visited[i-1][j] = true;
			}
		if ((i+1 <r)&&(!visited[i+1][j])&&(ch[i+1][j]=='.')){
			q.push({i+1,j});
			sp[{i+1,j}] = sp[{i,j}] + 1;
			visited[i+1][j]=true;
			}
		if ((j-1>=0)&&(!visited[i][j-1])&&(ch[i][j-1]=='.')){
			q.push({i,j-1});
			sp[{i,j-1}] = sp[{i,j}] + 1;
			visited[i][j-1]=true;
			}
		if ((j+1 <c)&&(!visited[i][j+1])&&(ch[i][j+1]=='.')){
			q.push({i,j+1});
			sp[{i,j+1}] = sp[{i,j}] + 1;
			visited[i][j+1]=true;
			}
		}
		int max = -1;
		tr(sp,it){
			int temp = (*it).second;
			if (temp>max){
				max = temp;
				}
			}
		return max;
	}


int main(){
	int t;
	cin>>t;
	
	while(t--){
		
	vector <int> ans;
	int c,r;
	cin>>c>>r;
	
	char ch[r+1][c+1];
	vector < vector <bool> > visited(1001, vector <bool> (1001, false));
	vector < vector <char> > cha;
	
	for(int i{0}; i<r; i++){
		cin>>ch[i];
		vector <char> t;
		int j=0;
		while (ch[i][j]){
			t.push_back(ch[i][j]);
			j++;
			}
		cha.push_back(t);
		}
		
	for(int i{0}; i<r; i++){
		for(int j{0}; j<c; j++){
			if (ch[i][j] == '.' && (!visited[i][j])){
				//cout<<"found . in ("<<i<<","<<j<<")"<<"\n";
				ans.push_back( bfs(i,j,r,c,visited,cha) );
				//cout<<"exit bfs"<<endl;
				break;
				}
			}
		}
		
	int max=-1;
	tr(ans,it){
		if ((*it)>max){
			max = (*it);
			//cout<<*it<<" ";
			}
		}
		cout<<"Maximum rope length is "<<max<<"."<<endl;
	
		
/*	for(int i{0}; i<r; i++){
		for(int j{0}; j<c; j++){
			cout<<ch[i][j]<<" ";
			}
			cout<<"\n";
		}
*/	
	visited.clear();
	ans.clear();
	cha.clear();
	}
	return 0;
	}
